src/codecs/webp/encoder.rs

```diff
@@ -0,0 +1,238 @@
+    Lossy(u8),
+}
+
+impl WebPQuality {
+    /// Minimum lossy quality value (0).
+    pub const MIN: u8 = 0;
+    /// Maximum lossy quality value (100).
+    pub const MAX: u8 = 100;
+    /// Default lossy quality (80), providing a balance of quality and file size.
+    pub const DEFAULT: u8 = 80;
+
+    /// Lossless encoding.
+    pub fn lossless() -> Self {
+        Self(Quality::Lossless)
+    }
+
+    /// Lossy encoding. 0 = low quality, small size; 100 = high quality, large size.
+    ///
+    /// Values are clamped from 0 to 100.
+    pub fn lossy(quality: u8) -> Self {
+        Self(Quality::Lossy(quality.clamp(Self::MIN, Self::MAX)))
+    }
+}
+
+impl Default for WebPQuality {
+    fn default() -> Self {
+        Self::lossy(WebPQuality::DEFAULT)
+    }
+}
+
+impl<W: Write> WebPEncoder<W> {
+    /// Create a new encoder that writes its output to `w`.
+    ///
+    /// Defaults to lossy encoding, see [`WebPQuality::DEFAULT`].
+    pub fn new(w: W) -> Self {
+        WebPEncoder::new_with_quality(w, WebPQuality::default())
+    }
+
+    /// Create a new encoder with the specified quality, that writes its output to `w`.
+    pub fn new_with_quality(w: W, quality: WebPQuality) -> Self {
+        Self { inner: w, quality }
+    }
+
+    /// Encode image data with the indicated color type.
+    ///
+    /// The encoder requires image data be Rgb8 or Rgba8.
+    pub fn encode(
+        mut self,
+        data: &[u8],
+        width: u32,
+        height: u32,
+        color: ColorType,
+    ) -> ImageResult<()> {
+        // TODO: convert color types internally?
+        let (layout, stride) = match color {
+            ColorType::Rgb8 => (PixelLayout::Rgb, 3),
+            ColorType::Rgba8 => (PixelLayout::Rgba, 4),
+            _ => {
+                return Err(ImageError::Unsupported(
+                    UnsupportedError::from_format_and_kind(
+                        ImageFormat::WebP.into(),
+                        UnsupportedErrorKind::Color(color.into()),
+                    ),
+                ))
+            }
+        };
+
+        // Validate dimensions upfront to avoid panics.
+        let expected_len = stride * (width * height) as u64;
+        if expected_len > data.len() as u64 {
+            return Err(ImageError::Parameter(ParameterError::from_kind(
+                ParameterErrorKind::DimensionMismatch,
+            )));
+        }
+
+        // Call the native libwebp library to encode the image.
+        let encoder = Encoder::new(data, layout, width, height);
+        let encoded: WebPMemory = match self.quality.0 {
+            Quality::Lossless => encoder.encode_lossless(),
+            Quality::Lossy(quality) => encoder.encode(quality as f32),
+        };
+
+        // The simple encoding API in libwebp does not return errors.
+        if encoded.is_empty() {
+            return Err(ImageError::Encoding(EncodingError::new(
+                ImageFormat::WebP.into(),
+                "encoding failed, output empty",
```

Comment:
Oh wow, I thought there was at least some error reporting but [no](https://github.com/webmproject/libwebp/blob/main/src/webp/encode.h#L43-L45). Okay, fine.

Reply:
So the "advanced WebP encoding API" does return errors, which the next release of the `webp` crate looks like it will include. When/if it's released, the implementation here can be swapped out internally to return errors (not that any are expected unless more configuration options are exposed from `image`) with no breaking changes.